services/checks/checks.go

```diff
@@ -419,6 +420,124 @@ func (s *Service) EnableChecks(checkNames []string) error {
 	return nil
 }
 
+// ChangeInterval changes a check's interval to the value received from the UI.
+func (s *Service) ChangeInterval(params map[string]managementpb.SecurityCheckInterval) error {
+	checks := s.GetAllChecks()
+	if len(checks) == 0 {
+		return errors.New("no checks loaded")
+	}
+
+	checkMap := make(map[string]check.Check, len(checks))
+	for _, ch := range checks {
+		checkMap[ch.Name] = ch
+	}
+
+	for name, interval := range params {
+		c, ok := checkMap[name]
+		if !ok {
+			return errors.Errorf("check: %s not found", name)
+		}
+
+		switch interval {
+		case managementpb.SecurityCheckInterval_STANDARD:
+			c.Interval = check.Standard
+		case managementpb.SecurityCheckInterval_FREQUENT:
+			c.Interval = check.Frequent
+		case managementpb.SecurityCheckInterval_RARE:
+			c.Interval = check.Rare
+		case managementpb.SecurityCheckInterval_SECURITY_CHECK_INTERVAL_INVALID:
+			return errors.New("invalid security check interval")
+		default:
+			return errors.New("unknown security check interval")
+		}
+
+		// since we re-run checks at regular intervals using a call
+		// to s.StartChecks which in turn calls s.collectChecks
+		// to load/download checks, we must persist any changes
+		// to check intervals in the DB so that they can be re-applied
+		// once the checks have been re-loaded on restarts.
+		e := s.db.InTransaction(func(tx *reform.TX) error {
+			cs, err := models.FindCheckStateByName(s.db.Querier, c.Name)
+			// record interval change for the first time.
+			if err == reform.ErrNoRows {
+				cs, err := models.CreateCheckState(s.db.Querier, c.Name, models.Interval(c.Interval))
+				if err != nil {
+					return err
+				}
+				s.l.Debugf("Saved interval change for check: %s in DB", cs.Name)
+				return nil
+			}
+
+			// update existing interval change.
+			if cs != nil {
+				cs, err := models.ChangeCheckState(s.db.Querier, c.Name, models.Interval(c.Interval))
+				if err != nil {
+					return err
+				}
+				s.l.Debugf("Updated interval change for check: %s in DB", cs.Name)
+				return nil
+			}
+			return err
+		})
+		if e != nil {
+			return e
+		}
+
+		s.updateCheck(c)
+	}
+
+	return nil
+}
+
+// updateCheck updates a check with an updated interval in the appropriate check slice.
+func (s *Service) updateCheck(newCheck check.Check) {
+	switch newCheck.Type {
+	case check.MySQLSelect:
+		fallthrough
+	case check.MySQLShow:
+		s.cm.Lock()
+		defer s.cm.Unlock()
+		for i, oldCheck := range s.mySQLChecks {
```

Comment:
How about making `s.mySQLChecks` map instead of slice? It will allow us to eliminate that loop.

Same for other checks slices.

Reply:
thats a good idea, removed a lot of unnecessary code thanks